### PR TITLE
Better error message when trying to write default impls

### PR DESCRIPTION
**Pull request by Phlosioneer**
Previously, if you tried to write this (using the specialization
feature flag):

default impl PartialEq<MyType> {
...
}

The compiler would give you the mysterious warning "inherent impls
cannot be default". What it really means is that you're trying to
write an impl for a Structure or *Trait Object*, and that cannot
be "default". However, one of the ways to encounter this error
(as shown by the above example) is when you forget to write "for
MyType".

This PR adds a help message that reads "maybe missing a `for`
keyword?" This is useful, actionable advice that will help any user
identify their mistake, and doesn't get in the way or mislead any
user that really meant to use the "default" keyword for this weird
purpose. In particular, this help message will be useful for any
users who don't know the "inherent impl" terminology, and/or users
who forget that inherent impls CAN be written for traits (they apply
to the trait objects). Both of these are somewhat confusing, seldom-
used concepts; a one-line error message without any error number for
longer explanation is NOT the place to introduce these ideas.

I wasn't quite sure what grammar / wording to use. I'm open to suggestions. CC @rust-lang/docs (I hope I'm doing that notation right)

(Apparently not. :( )

### CODE REVIEW

**Comment by pietroalbini** [comment]
Highfive failed to pick a reviewer for this PR.

r? @estebank

File: src/librustc_passes/ast_validation.rs [hunk 1]
```diff
@@ -272,7 +272,9 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                     self.err_handler().span_err(item.span, "inherent impls cannot be negative");
                 }
                 if defaultness == Defaultness::Default {
-                    self.err_handler().span_err(item.span, "inherent impls cannot be default");
+                    self.err_handler()
+                        .struct_span_err(item.span, "inherent impls cannot be default")
+                        .help("maybe a missing `for` keyword?");
```
Comment (estebank) [inline src/librustc_passes/ast_validation.rs]: This is creating a `DiagnosticBuilder` but never has `.emit()` called on it, which causes an Internal Compiler Error in test `compile-fail/specialization/defaultimpl/validation.rs`.

Assign the result of the `struct_span_err` to a mutable variable `err` and then `err.help(...)` and `err.emit()`.

**Review by estebank — CHANGES_REQUESTED** [review]
I would prefer a suggestion if it could be provided, with text close to

    help: if you meant to write an `impl` for `PartialEq`, use the `for` keyword: `default impl for PartialEq`

> (I hope I'm doing that notation right)

Only members of the org can CC teams.

CC @rust-lang/docs 

I'm gonna be offline for two weeks. Apologies in advance for the unresponsiveness.

**Comment by petrochenkov** [comment]
>one of the ways to encounter this error
(as shown by the above example) is when you forget to write "for
MyType".

And another way is to actually write `default impl MyStruct<T> { ... }` in attempt to use specialization on inherent methods.
You can't recommend `for` without checking that `X` in `default impl X { ... }` is actually a trait.

File: src/librustc_passes/ast_validation.rs (hunk 1 above)
Comment (Phlosioneer) [reply src/librustc_passes/ast_validation.rs]: Oh, okay. Sorry, I'm not very familiar with internals. And there's no rush for feedback, thanks for the warning.

**Comment by Phlosioneer** [comment]
True... I don't know that we can do that here, though. Isn't this AST validation, before we've resolved trait names and the like?

Also, I don't know how to do that; I'm not very familiar with the rustc internals. Do you happen to know how to do that, or someplace I can start looking?

**Comment by shepmaster** [comment]
Since @estebank is out, randomly assigning to...

r? @nagisa

**Comment by pietroalbini** [comment]
Ping from triage @Phlosioneer! The reviewer left some comments, could you address them?

**Comment by pietroalbini** [comment]
Ping from triage @nagisa! The author addressed your concerns, could you review this again?

**Pull request merged**